src/core/common/message.hpp

```diff
@@ -65,6 +63,61 @@ class ThreadLinkInfo;
  *
  */
 
+/**
+ * This macro frees a given message buffer if not nullptr.
+ *
+ * This macro and the ones that follow contain small but common code patterns used in many of the core modules. They
+ * are intentionally defined as macros instead of inline methods/functions to ensure that they are fully inlined.
+ * Note that an `inline` method/function is not necessarily always inlined by the toolchain and not inlining such
+ * small implementations can add a rather large code-size overhead.
+ *
+ * @param[in] aMessage    A pointer to a `Message` to free (can be nullptr).
+ *
+ */
+#define FreeMessage(aMessage)      \
+    do                             \
+    {                              \
+        if ((aMessage) != nullptr) \
+        {                          \
+            (aMessage)->Free();    \
+        }                          \
+    } while (false)
+
+/**
+ * This macro frees a given message buffer if a given `otError` indicates an error.
+ *
+ * The parameter @p aMessage can be nullptr in which case this macro does nothing.
+ *
+ * @param[in] aMessage    A pointer to a `Message` to free (can be nullptr).
+ * @param[in] aError      The `otError` to check.
+ *
+ */
+#define FreeMessageOnError(aMessage, aError)                        \
+    do                                                              \
+    {                                                               \
+        if (((aError) != OT_ERROR_NONE) && ((aMessage) != nullptr)) \
+        {                                                           \
+            (aMessage)->Free();                                     \
```

Comment:
```suggestion
            (aMessage)->Free();                                     \
            (aMessage) = nullptr;                                    \
```

If this does not increase code size, maybe we can avoid defining `FreeAndNullMessageOnError` ?

Reply:
We cannot use it. It does impact the code size.

Reply:
I tested it with nrf52840 build and it didn't increase code size. But not sure if all compilers can do such optimization (dead code elimination).

Reply:
I know some toolchain could optimize the unused assignment and remove the code but we would then need to rely on compiler/toolchain behavior/optimization. 
My recommendation/approach is not to rely on toolchain behavior. Better be explicit and clear when/if we can.

Reply:
SGTM